include/swift/Runtime/Mutex.h

```diff
@@ -186,121 +114,593 @@ class Mutex {
+  ///   If ownership is not obtained, returns immediately. The function is
+  ///   allowed to spuriously fail and return even if the lock is not
+  ///   currently owned by another thread.
+  /// - If try_readLock() succeeds, prior unlock() operations on the same
+  ///   object synchronize-with this operation. unlock() does not synchronize
+  ///   with a failed try_readLock().
+  /// - The behavior is undefined if the calling thread already owns
+  ///   the read or write lock (likely a deadlock)?
+  /// - Does not throw exceptions but will halt on error (fatalError).
+  ///
+  /// Callers must not mutate the data protected by the ReadWriteLock while
+  /// holding the read lock, the write lock must be used.
+  bool try_readLock() {
+    return ReadWriteLockPlatformHelper::try_readLock(Handle);
+  }
+
+  /// The writeLock() method has the following properties:
+  /// - Behaves as an atomic operation.
+  /// - Blocks the calling thread while the write lock or a read lock is held
+  ///   by another thread and once the write lock is acquired by the calling
+  ///   thread other threads are prevented from acquiring the write lock or a
+  ///   read lock.
+  /// - Only one thread can hold the write lock at the same time.
+  /// - Prior unlock() operations on the same lock synchronize-with
+  ///   this lock operation.
+  /// - The behavior is undefined if the calling thread already owns
+  ///   the read or write lock (likely a deadlock).
+  /// - Does not throw exceptions but will halt on error (fatalError).
+  void writeLock() { ReadWriteLockPlatformHelper::writeLock(Handle); }
+
+  /// The try_writeLock() method has the following properties:
+  /// - Behaves as an atomic operation.
+  /// - Attempts to obtain the write lock without blocking the calling thread.
+  ///   If ownership is not obtained, returns immediately. The function is
+  ///   allowed to spuriously fail and return even if the lock is not
+  ///   currently owned by another thread.
+  /// - If try_writeLock() succeeds, prior unlock() operations on the same
+  ///   object synchronize-with this operation. unlock() does not synchronize
+  ///   with a failed try_writeLock().
+  /// - The behavior is undefined if the calling thread already owns
+  ///   the read or write lock (likely a deadlock)?
+  /// - Does not throw exceptions but will halt on error (fatalError).
+  bool try_writeLock() {
+    return ReadWriteLockPlatformHelper::try_writeLock(Handle);
+  }
+
+  /// The unlock() method has the following properties:
+  /// - Behaves as an atomic operation.
+  /// - Releases the calling thread's ownership of the read or write lock
+  ///   and synchronizes-with the subsequent successful lock operations on
+  ///   the same object.
+  /// - The behavior is undefined if the calling thread does not own
+  ///   either the read or write lock.
+  /// - Does not throw exceptions but will halt on error (fatalError).
+  void unlock() { ReadWriteLockPlatformHelper::unlock(Handle); }
+
```

Comment:
@tinysun212 glad you are looking into a Win32 version of this, sorry for changing things around some form what you likely first saw, decided to rework things to hopefully better support the variants that are potentially needed. Let me know if you hit any issues or have questions.